File: android/src/main/kotlin/com/capacitorjs/plugins/calendar/CalendarPlugin.kt
package com.capacitorjs.plugins.calendar

import android.Manifest
import android.content.ActivityNotFoundException
import android.content.ContentUris
import android.content.Intent
import android.provider.CalendarContract
import android.provider.CalendarContract.Events
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * Capacitor bridge for the Calendar plugin. Parses calls, ensures the needed
 * permissions, and delegates to [Calendar]. Provider work runs
 * on a single-thread executor; the interactive editor and the calendar app
 * run through activity intents.
 */
@CapacitorPlugin(
    name = "Calendar",
    permissions = [
        Permission(strings = [Manifest.permission.READ_CALENDAR], alias = CalendarPlugin.READ),
        Permission(strings = [Manifest.permission.WRITE_CALENDAR], alias = CalendarPlugin.WRITE)
    ]
)
class CalendarPlugin : Plugin() {

    companion object {
        const val READ = "readCalendar"
        const val WRITE = "writeCalendar"
    }

    private val implementation by lazy { Calendar(context) }
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()

    /** `true` while the interactive event editor is open. */
    @Volatile
    private var editInProgress = false

    override fun handleOnDestroy() {
        executor.shutdown()
    }

    // ------------------------------------------------------------------
    // Events
    // ------------------------------------------------------------------

    @PluginMethod
    fun createEvent(call: PluginCall) {
        withPermissions(call, "createEventPermissionCallback", WRITE) {
            run(call) { call.resolve(JSObject().put("id", implementation.createEvent(call.data))) }
        }
    }

    @PermissionCallback
    private fun createEventPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, WRITE) {
            run(call) { call.resolve(JSObject().put("id", implementation.createEvent(call.data))) }
        }
    }

    @PluginMethod
    fun createEventInteractively(call: PluginCall) {
        if (editInProgress) {
            reject(call, CalendarError.PENDING_OPERATION)
            return
        }
        val title = call.getString("title")
        val startMs = call.getLong("startDate")
        val endMs = call.getLong("endDate")
        if (title.isNullOrEmpty() || startMs == null || endMs == null || endMs < startMs) {
            reject(call, CalendarError.INVALID_ARGUMENT)
            return
        }

        val intent = Intent(Intent.ACTION_INSERT)
            .setData(Events.CONTENT_URI)
            .putExtra(Events.TITLE, title)
            .putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, startMs)
            .putExtra(CalendarContract.EXTRA_EVENT_END_TIME, endMs)
            .putExtra(CalendarContract.EXTRA_EVENT_ALL_DAY, call.getBoolean("isAllDay") == true)
        call.getString("location")?.let { intent.putExtra(Events.EVENT_LOCATION, it) }
        call.getString("notes")?.let { intent.putExtra(Events.DESCRIPTION, it) }

        editInProgress = true
        try {
            startActivityForResult(call, intent, "editEventResult")
        } catch (e: ActivityNotFoundException) {
            editInProgress = false
            reject(call, CalendarError.NOT_SUPPORTED)
        }
    }

    @ActivityCallback
    private fun editEventResult(call: PluginCall?, result: ActivityResult) {
        editInProgress = false
        // The calendar app reports no result data and no reliable result
        // code, so a saved event's id (and a cancel) cannot be detected.
        call?.resolve(JSObject())
    }

    @PluginMethod
    fun modifyEvent(call: PluginCall) {
        val filter = call.getObject("filter")
        val newEvent = call.getObject("newEvent")
        if (filter == null || newEvent == null) {
            reject(call, CalendarError.INVALID_ARGUMENT)
            return
        }
        withPermissions(call, "modifyEventPermissionCallback", READ, WRITE) {
            run(call) {
                implementation.modifyEvent(filter, newEvent)
                call.resolve()
            }
        }
    }

    @PermissionCallback
    private fun modifyEventPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, READ, WRITE) {
            run(call) {
                implementation.modifyEvent(call.getObject("filter")!!, call.getObject("newEvent")!!)
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun findEvents(call: PluginCall) {
        withPermissions(call, "findEventsPermissionCallback", READ) {
            run(call) { call.resolve(JSObject().put("events", implementation.findEvents(call.data))) }
        }
    }

    @PermissionCallback
    private fun findEventsPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, READ) {
            run(call) { call.resolve(JSObject().put("events", implementation.findEvents(call.data))) }
        }
    }

    @PluginMethod
    fun deleteEvent(call: PluginCall) {
        val hasId = !call.getString("id").isNullOrEmpty()
        val hasFilter = listOf("title", "location", "notes", "startDate", "endDate", "calendarName")
            .any { call.data.has(it) }
        if (!hasId && !hasFilter) {
            reject(call, CalendarError.INVALID_ARGUMENT)
            return
        }
        withPermissions(call, "deleteEventPermissionCallback", READ, WRITE) {
            run(call) {
                implementation.deleteEvent(call.data)
                call.resolve()
            }
        }
    }

    @PermissionCallback
    private fun deleteEventPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, READ, WRITE) {
            run(call) {
                implementation.deleteEvent(call.data)
                call.resolve()
            }
        }
    }

    // ------------------------------------------------------------------
    // Calendars
    // ------------------------------------------------------------------

    @PluginMethod
    fun listCalendars(call: PluginCall) {
        withPermissions(call, "listCalendarsPermissionCallback", READ) {
            run(call) { call.resolve(JSObject().put("calendars", implementation.listCalendars())) }
        }
    }

    @PermissionCallback
    private fun listCalendarsPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, READ) {
            run(call) { call.resolve(JSObject().put("calendars", implementation.listCalendars())) }
        }
    }

    @PluginMethod
    fun createCalendar(call: PluginCall) {
        val name = call.getString("name")
        if (name.isNullOrEmpty()) {
            reject(call, CalendarError.INVALID_ARGUMENT)
            return
        }
        withPermissions(call, "createCalendarPermissionCallback", WRITE) {
            run(call) {
                call.resolve(JSObject().put("id", implementation.createCalendar(name, call.getString("color"))))
            }
        }
    }

    @PermissionCallback
    private fun createCalendarPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, WRITE) {
            run(call) {
                call.resolve(
                    JSObject().put("id", implementation.createCalendar(call.getString("name")!!, call.getString("color")))
                )
            }
        }
    }

    @PluginMethod
    fun deleteCalendar(call: PluginCall) {
        val name = call.getString("name")
        if (name.isNullOrEmpty()) {
            reject(call, CalendarError.INVALID_ARGUMENT)
            return
        }
        withPermissions(call, "deleteCalendarPermissionCallback", READ, WRITE) {
            run(call) {
                implementation.deleteCalendar(name)
                call.resolve()
            }
        }
    }

    @PermissionCallback
    private fun deleteCalendarPermissionCallback(call: PluginCall) {
        continueWithPermissions(call, READ, WRITE) {
            run(call) {
                implementation.deleteCalendar(call.getString("name")!!)
                call.resolve()
            }
        }
    }

    @PluginMethod
    fun openCalendar(call: PluginCall) {
        val ms = call.getLong("date") ?: System.currentTimeMillis()
        val uri = ContentUris.appendId(
            CalendarContract.CONTENT_URI.buildUpon().appendPath("time"),
            ms
        ).build()
        try {
            activity.startActivity(Intent(Intent.ACTION_VIEW).setData(uri))
            call.resolve()
        } catch (e: ActivityNotFoundException) {
            reject(call, CalendarError.NOT_SUPPORTED)
        }
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private fun hasPermissions(vararg aliases: String): Boolean =
        aliases.all { getPermissionState(it) == com.getcapacitor.PermissionState.GRANTED }

    /**
     * Runs `body` when the aliases are granted, otherwise requests them and
     * resumes in the named [PermissionCallback].
     */
    private fun withPermissions(call: PluginCall, callbackName: String, vararg aliases: String, body: () -> Unit) {
        if (hasPermissions(*aliases)) {
            body()
        } else {
            requestPermissionForAliases(aliases, call, callbackName)
        }
    }

    private fun continueWithPermissions(call: PluginCall, vararg aliases: String, body: () -> Unit) {
        if (hasPermissions(*aliases)) {
            body()
        } else {
            reject(call, CalendarError.PERMISSION_DENIED)
        }
    }

    /** Runs `body` on the provider executor, mapping thrown errors to rejections. */
    private fun run(call: PluginCall, body: () -> Unit) {
        executor.execute {
            try {
                body()
            } catch (e: CalendarException) {
                reject(call, e.error)
            } catch (e: SecurityException) {
                reject(call, CalendarError.PERMISSION_DENIED)
            } catch (e: Exception) {
                reject(call, CalendarError.UNKNOWN)
            }
        }
    }

    private fun reject(call: PluginCall, error: CalendarError) {
        call.reject(error.message, error.code)
    }
}

File: ios/Sources/CalendarPlugin/CalendarPlugin.swift
import Capacitor
import EventKit
import EventKitUI
import Foundation

/// Capacitor bridge for the Calendar plugin. Parses calls, ensures the
/// needed access level, and delegates to `Calendar`.
@objc(CalendarPlugin)
public class CalendarPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CalendarPlugin"
    public let jsName = "Calendar"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createEventInteractively", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "modifyEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "findEvents", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listCalendars", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createCalendar", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteCalendar", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openCalendar", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = Calendar()

    /// Call awaiting the interactive event editor.
    private var editCall: CAPPluginCall?
    private var editDelegate: EditDelegate?

    // MARK: - Permissions

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        let states = implementation.permissionStates()
        call.resolve(["readCalendar": states.read, "writeCalendar": states.write])
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        let requested = call.getArray("permissions", String.self) ?? ["readCalendar", "writeCalendar"]
        implementation.requestAccess(read: requested.contains("readCalendar")) { [weak self] in
            self?.checkPermissions(call)
        }
    }

    // MARK: - Events

    @objc func createEvent(_ call: CAPPluginCall) {
        withAccess(call, read: false) {
            do {
                let id = try self.implementation.createEvent(self.options(of: call))
                call.resolve(["id": id])
            } catch {
                self.reject(call, error)
            }
        }
    }

    @objc func createEventInteractively(_ call: CAPPluginCall) {
        guard editCall == nil else {
            reject(call, CalendarError.pendingOperation)
            return
        }
        // iOS 17+ shows the editor without calendar permission; earlier
        // versions need access before the store can save from the editor.
        if #available(iOS 17.0, *) {
            presentEditor(call)
        } else {
            withAccess(call, read: false) {
                self.presentEditor(call)
            }
        }
    }

    @objc func modifyEvent(_ call: CAPPluginCall) {
        guard let filter = call.getObject("filter"), let newEvent = call.getObject("newEvent") else {
            reject(call, CalendarError.invalidArgument)
            return
        }
        withAccess(call, read: true) {
            do {
                try self.implementation.modifyEvent(filter: filter, newEvent: newEvent)
                call.resolve()
            } catch {
                self.reject(call, error)
            }
        }
    }

    @objc func findEvents(_ call: CAPPluginCall) {
        withAccess(call, read: true) {
            do {
                let events = try self.implementation.findEvents(self.options(of: call))
                call.resolve(["events": events])
            } catch {
                self.reject(call, error)
            }
        }
    }

    @objc func deleteEvent(_ call: CAPPluginCall) {
        let hasId = (call.getString("id")?.isEmpty == false)
        let hasFilter = ["title", "location", "notes", "startDate", "endDate", "calendarName"]
            .contains { call.options?[$0] != nil }
        guard hasId || hasFilter else {
            reject(call, CalendarError.invalidArgument)
            return
        }
        withAccess(call, read: true) {
            do {
                try self.implementation.deleteEvent(self.options(of: call))
                call.resolve()
            } catch {
                self.reject(call, error)
            }
        }
    }

    // MARK: - Calendars

    @objc func listCalendars(_ call: CAPPluginCall) {
        withAccess(call, read: true) {
            call.resolve(["calendars": self.implementation.listCalendars()])
        }
    }

    @objc func createCalendar(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), !name.isEmpty else {
            reject(call, CalendarError.invalidArgument)
            return
        }
        withAccess(call, read: false) {
            do {
                let id = try self.implementation.createCalendar(name: name, color: call.getString("color"))
                call.resolve(["id": id])
            } catch {
                self.reject(call, error)
            }
        }
    }

    @objc func deleteCalendar(_ call: CAPPluginCall) {
        guard let name = call.getString("name"), !name.isEmpty else {
            reject(call, CalendarError.invalidArgument)
            return
        }
        withAccess(call, read: true) {
            do {
                try self.implementation.deleteCalendar(name: name)
                call.resolve()
            } catch {
                self.reject(call, error)
            }
        }
    }

    @objc func openCalendar(_ call: CAPPluginCall) {
        let ms = call.getDouble("date") ?? Date().timeIntervalSince1970 * 1000.0
        let date = Date(timeIntervalSince1970: ms / 1000.0)
        let seconds = Int(date.timeIntervalSinceReferenceDate)
        DispatchQueue.main.async {
            guard let url = URL(string: "calshow:\(seconds)") else {
                self.reject(call, CalendarError.unknown)
                return
            }
            UIApplication.shared.open(url, options: [:]) { opened in
                if opened {
                    call.resolve()
                } else {
                    self.reject(call, CalendarError.notSupported)
                }
            }
        }
    }

    // MARK: - Helpers

    /// Runs `body` on a background queue once the needed access level is
    /// granted, requesting it first when not yet determined.
    private func withAccess(_ call: CAPPluginCall, read: Bool, _ body: @escaping () -> Void) {
        let granted = read ? implementation.hasReadAccess() : implementation.hasWriteAccess()
        if granted {
            DispatchQueue.global(qos: .userInitiated).async(execute: body)
            return
        }
        implementation.requestAccess(read: read) { [weak self] in
            guard let self = self else { return }
            let granted = read ? self.implementation.hasReadAccess() : self.implementation.hasWriteAccess()
            if granted {
                DispatchQueue.global(qos: .userInitiated).async(execute: body)
            } else {
                self.reject(call, CalendarError.permissionDenied)
            }
        }
    }

    private func presentEditor(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let event = EKEvent(eventStore: self.implementation.store)
            do {
                try self.implementation.apply(self.options(of: call), to: event, requireDates: true)
            } catch {
                self.reject(call, error)
                return
            }

            guard var presenter = self.bridge?.viewController else {
                self.reject(call, CalendarError.unknown)
                return
            }
            while let presented = presenter.presentedViewController {
                presenter = presented
            }

            self.editCall = call
            let delegate = EditDelegate { [weak self] action, savedId in
                guard let self = self, let call = self.editCall else { return }
                self.editCall = nil
                self.editDelegate = nil
                if action == .saved {
                    var result: [String: Any] = [:]
                    if let savedId = savedId {
                        result["id"] = savedId
                    }
                    call.resolve(result)
                } else {
                    self.reject(call, CalendarError.operationCancelled)
                }
            }
            self.editDelegate = delegate

            let controller = EKEventEditViewController()
            controller.eventStore = self.implementation.store
            controller.event = event
            controller.editViewDelegate = delegate
            presenter.present(controller, animated: true)
        }
    }

    private func options(of call: CAPPluginCall) -> [String: Any] {
        return (call.options as? [String: Any]) ?? [:]
    }

    private func reject(_ call: CAPPluginCall, _ error: Error) {
        let calendarError = error as? CalendarError ?? .unknown
        call.reject(calendarError.message, calendarError.code)
    }
}

/// `EKEventEditViewDelegate` wrapper reporting the edit outcome and, on save,
/// the created event's id.
private class EditDelegate: NSObject, EKEventEditViewDelegate {
    private let completion: (EKEventEditViewAction, String?) -> Void

    init(completion: @escaping (EKEventEditViewAction, String?) -> Void) {
        self.completion = completion
    }

    func eventEditViewController(_ controller: EKEventEditViewController, didCompleteWith action: EKEventEditViewAction) {
        let savedId = action == .saved ? controller.event?.calendarItemIdentifier : nil
        controller.dismiss(animated: true) {
            self.completion(action, savedId)
        }
    }
}
